src/ray/util/shared_lru.h

```diff
@@ -0,0 +1,206 @@
+// Copyright 2024 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SharedLruCache is a LRU cache, with all entries shared, which means a single entry
+// could be accessed by multiple getters. All values are wrapped with shared pointer to
+// avoid copy at get operation, meanwhile also useful to maintain memory validity at any
+// time.
+//
+// Example usage:
+// SharedLruCache<std::string, std::string> cache{cap};
+// // Put a key-value pair into cache.
+// cache.Put("key", std::make_shared<std::string>("val"));
+//
+// // Get a key-value pair from cache.
+// auto val = cache.Get("key");
+// // Check and consume `val`.
+//
+// TODO(hjiang):
+// 1. Add template arguments for key hash and key equal, to pass into absl::flat_hash_map.
+// 2. Provide a key hash wrapper to save a copy.
+// 3. flat hash map supports heterogeneous lookup, expose `KeyLike` templated interface.
+// 4. Add a `GetOrCreate` interface, which takes factory function to creation value.
+// 5. For thread-safe cache, add a sharded container wrapper to reduce lock contention.
+
+#pragma once
+
+#include <cstdint>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <optional>
+#include <string>
+#include <utility>
+
+#include "absl/container/flat_hash_map.h"
+#include "src/ray/util/logging.h"
+
+namespace ray::utils::container {
+
+template <typename Key, typename Val>
+class SharedLruCache final {
+ public:
+  using key_type = Key;
+  using mapped_type = Val;
+
+  // A `max_entries` of 0 means that there is no limit on the number of entries
+  // in the cache.
+  explicit SharedLruCache(size_t max_entries) : max_entries_(max_entries) {}
+
+  SharedLruCache(const SharedLruCache &) = delete;
+  SharedLruCache &operator=(const SharedLruCache &) = delete;
+
+  ~SharedLruCache() = default;
+
+  // Insert `value` with key `key`. This will replace any previous entry with
+  // the same key.
+  void Put(Key key, std::shared_ptr<Val> value) {
+    RAY_CHECK(value != nullptr);
+    auto iter = cache_.find(key);
+    if (iter != cache_.end()) {
+      lru_list_.splice(lru_list_.begin(), lru_list_, iter->second.lru_iterator);
+      iter->second.value = std::move(value);
+    } else {
+      lru_list_.emplace_front(key);
+      Entry new_entry{std::move(value), lru_list_.begin()};
+      cache_[std::move(key)] = std::move(new_entry);
+    }
+
+    if (max_entries_ > 0 && lru_list_.size() > max_entries_) {
```

Comment:
this eviction can only happen in the `else` case, that is, `iter != cache_.end()`. put it there.

Reply:
Updated.